src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs

```diff
@@ -73,22 +69,19 @@ public static object Deserialize(string json, Type returnType, JsonSerializerOpt
             object result;
             byte[] tempArray = null;
 
-            int maxBytes;
-
-            // For performance, avoid asking for the actual byte count unless necessary.
-            if (json.Length > MaxArrayLengthBeforeCalculatingSize)
+            // For performance, avoid obtaining the actual byte count unless the memory usage may be
+            // higher than the threshold which is options.DefaultBufferSize.
+            long maxBytes = (long)json.Length * JsonConstants.MaxExpansionFactorWhileTranscoding;
+            if (maxBytes > options.DefaultBufferSize)
             {
                 // Get the actual byte count in order to handle large input.
                 maxBytes = JsonReaderHelper.GetUtf8ByteCount(json.AsSpan());
-            }
-            else
-            {
-                maxBytes = json.Length * JsonConstants.MaxExpansionFactorWhileTranscoding;
+                Debug.Assert(maxBytes <= int.MaxValue);
             }
 
             Span<byte> utf8 = maxBytes <= JsonConstants.StackallocThreshold ?
-                stackalloc byte[maxBytes] :
-                (tempArray = ArrayPool<byte>.Shared.Rent(maxBytes));
+                stackalloc byte[(int)maxBytes] :
```

Comment:
With the 3x multiplication, is it really worth it to stackalloc? It looks like we'd only stackalloc for strings of < 86 characters. Is that useful? Maybe do 3 * StackallocThreshold?

Reply:
See above; removing